add (experimental) lightweight/basic bandwidth telemetry to fabric.

SeanNijjar opened the pull request:
This PR adds an opt-in, experimental bandwidth telemetry feature to TT-Fabric router. The approach here is minimal and future work will provide more resolution/granularity to the reporting.

The current BW telemetry behaviour is reported once for the entire workload. It works by timestamping windows where we either send messages or have unsent/unacknowledged messages in our sender channel. For windows where no traffic is recorded, the interval is omitted. This is to avoid reported bandwidth artificially being lowered due to long idle times in the test.

Results are currently dumped in L1 at the first unreserved address (this is subject to change) and take the following structure (also subject to change):

```
struct results {
  uint64_t elapsed_cycles;
  uint64_t zero;
  // eth words are 16B, this lets us have longer capture windows without needing a uin64_t to manage overflow
  uint32_t num_eth_words_sent; 
  uint32_t num_packets_sent;
}
```

Future iterations will add more functionality:
- Finer grain telemetry
  - Host polling rate dictates resolution of these bandwidth measurements, as opposed to one number for a full run
  - Will move it closer to "real-time" reporting from the  
- Per (sender/receiver) channel reporting

## Usage

To enable the telemetry, specifiy environment variable `TT_METAL_FABRIC_TELEMETRY=1`. This flag will only enable the telemetry on device side, and will not (currently) enable automatic reporting by host of telemetry data.

As a rough prototype, the fabric bandwidth tests have implemented a basic read-back function.

## Limitation
Because this current implementation strictly accumulates into a fixed set of counters, with no ability to swap out data, there is an integer overflow concern. Because of this, results are only considered valid if the test runtime is less than about 10s of runtime. This is fine for our intended uses cases. This will be lifted in the future as we enable multiple buffer slots for results that can be round-robin read by host and cycled through by the router.

## Overheads

Overhead data is still being collected but ranges depending on the fabric and type of traffic pattern. So far I have observed a 0-1% overhead. However, further testing may reveal higher max overheads. This is another reason the feature is opt-in for the time-being.

## Outcomes

Based on the existing telemetry in this change, we are already able to start plotting useful information for benchmarking and system-level performance analysis.

For example, this is a line fabric running on the inner 4 devices of a t3k with traffic being sent in both directions. We see that the middle hop it "hot" and the end hops are cooler (about half bandwidth), due to the current fairness policy implemented in the router.

<img width="1309" height="770" alt="image" src="https://github.com/user-attachments/assets/187387ab-b8c0-4a92-9598-df3ec1ab1e5e" />

Here's another example with a full T3K where all rows and columns are sending data left/right/up/down. Note that the labels/lines may not exactly line up - the image is really more for illustrative purposes. Also note that the labels specify "Gbps" but the actual unit of the numbers are  "GB/s". This issue is resolved in this set of changes.
<img width="1394" height="828" alt="image" src="https://github.com/user-attachments/assets/c2147218-3ad9-40ca-9824-6a0e52f08e2b" />

The images above are not intended to represent a productized visualizer but are more of a quick visualization aid. They are generated by the `visualize_bw.py` script in this change.


### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI: https://github.com/tenstorrent/tt-metal/actions/runs/16686836591
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable): https://github.com/tenstorrent/tt-metal/actions/runs/16506442679
- [x] Microbenchmark: https://github.com/tenstorrent/tt-metal/actions/runs/16506444544
- [x] New tests in t3k unit tests: https://github.com/tenstorrent/tt-metal/actions/runs/16510001431

tt-aho commented on tt_metal/fabric/hw/inc/edm_fabric/telemetry/fabric_bandwidth_telemetry.hpp:
```diff
@@ -0,0 +1,271 @@
+#include "tt_metal/hw/inc/risc_attribs.h"
+
+#include <cstdint>
+#include <cstddef>  // Define FORCE_INLINE if not already defined
+#ifndef FORCE_INLINE
+#define FORCE_INLINE __attribute__((always_inline)) inline
+#endif
```
Minor: Why is this needed?

SeanNijjar replied:
it's not. I'll include a header that defines this. Side note: I think fabric will probably need to define its own variant of this macro so we have more control of when we do/don't apply this attribute (e.g. in 2D vs 1D or BH vs WH, we want different inlining behaviour in some cases).

tt-aho commented on tests/ttnn/unit_tests/gtests/ccl/test_fabric_edm_common.hpp:
```diff
@@ -2723,6 +2723,233 @@ tt::tt_fabric::FabricRouterBufferConfig get_edm_buffer_config_helper(FabricTestM
+                "Clearing telemetry data for device {}, ethernet core {}",
+                device->id(),
+                eth_logical_core);
+            // tt::tt_metal::detail::WriteToDeviceL1(
+            //     device, eth_logical_core, telemetry_buffer_address, clearing_data_vec, CoreType::ETH);
```
Is this function still needed since it's a no-op?

SeanNijjar replied:
thanks, it's deadcode that I missed.

tt-aho commented on tt_metal/fabric/erisc_datamover_builder.cpp:
```diff
@@ -149,6 +149,11 @@ FabricEriscDatamoverConfig::FabricEriscDatamoverConfig(Topology topology) {
     // Global
     size_t next_l1_addr = tt::tt_metal::hal::get_erisc_l1_unreserved_base();
+    if (tt::tt_metal::MetalContext::instance().rtoptions().get_enable_fabric_telemetry()) {
+        this->perf_telemetry_buffer_address = next_l1_addr;
+        next_l1_addr += 32;
```
Is there an assert we can add that this size is enough?

SeanNijjar replied:
I'll add an assert somewhere for this

tt-aho commented on tests/ttnn/unit_tests/gtests/ccl/test_fabric_edm_common.hpp:
```diff
@@ -2723,6 +2723,233 @@ tt::tt_fabric::FabricRouterBufferConfig get_edm_buffer_config_helper(FabricTestM
 
+// Clears the L1 buffer on every active ethernet core where telemetry data is buffered
+void clear_fabric_telemetry_data(MeshDeviceView* mesh_device) {
+    constexpr size_t telemetry_struct_size_bytes = 24;
+    const auto telemetry_buffer_address = tt::tt_metal::hal::get_erisc_l1_unreserved_base();
```
Minor: I guess we don't have a clean way of querying these values, which is why they're hardcoded in these tests?

SeanNijjar replied:
For the time being, yes. I'm going to sync with @btrzynadlowski-tt about how we'd like to structure this data going forward. It's a bit adhoc at the moment (also why I've relegated this read-back to only test code for the time being).

yugaoTT reviewed (approved):
LGTM

SeanNijjar commented on tt_metal/fabric/hw/inc/edm_fabric/telemetry/fabric_bandwidth_telemetry.hpp:
```diff
@@ -0,0 +1,271 @@
+FORCE_INLINE void write_perf_recording_window_results<LowResolutionBandwidthTelemetry, L1PerfTelemetrySingleBuffer>(
+    LowResolutionBandwidthTelemetry& perf_telemetry_collector, L1PerfTelemetrySingleBuffer& buffer) {
+    volatile tt_l1_ptr auto* buffer_ptr = buffer.get_ptr();
+    auto num_words_sent_l1_copy = buffer_ptr[4];
+    auto num_packets_sent_l1_copy = buffer_ptr[5];
+    uint64_t total_cycles = (static_cast<uint64_t>(buffer_ptr[1]) << 32) | buffer_ptr[0];
+    auto added_cycles =
+        total_cycles + (perf_telemetry_collector.timestamp_end.full - perf_telemetry_collector.timestamp_start.full);
+
+    // For the time being, this telemetry mode records elapsed elapsed (busy) cycles, not start/end, because
+    // we want to exclude large idle times from the bandwidth calculation
+    buffer_ptr[0] = added_cycles & 0xFFFFFFFF;
+    buffer_ptr[1] = added_cycles >> 32;
+    // Skip end timstamp
+    buffer_ptr[4] = num_words_sent_l1_copy + perf_telemetry_collector.num_words_sent;
+    buffer_ptr[5] = num_packets_sent_l1_copy + perf_telemetry_collector.num_packets_sent;
```
How do you mean? My intention here is that we'll define separate `write_perf_recording_window_results` definitions for the various telemetry recorder types (in this one, there is just the one, or none as options). 

I want us to be able to pick and choose the level of telemetry we have here, so I want to avoid having a global impl and instead would prefer discrete implementations for the different levels of telemetry. A more granular BW telemetry (e.g. per sender channel) would have a different impl of this function.

In terms of toggling different types of telemetry (BW vs packet trace vs others) in any combination, I think this approach is still compatible, but I'm definitely flexible on us changing the impl. I tried to make this as non-committal of an impl as I could until we land on how we want to formally structure all this stuff.

yugaoTT replied:
or like this, 
```

struct TelemetryBuffer {
  uint64_t elapsed_cycles;       // words [0]+[1]
  uint64_t skip_timestamp;       // words [2]+[3] — unused here
  uint32_t num_eth_words_sent;   // word  [4]
  uint32_t num_packets_sent;     // word  [5]
};
```
then 
```
uint64_t prev_words = buf->num_eth_words_sent;
uint64_t prev_pkts = buf->num_packets_sent;

// compute new cycles:
buf->elapsed_cycles +=
    (perf_telemetry_collector.timestamp_end.full
   - perf_telemetry_collector.timestamp_start.full);
```

aagarwalTT commented on tests/scripts/t3000/run_t3000_unit_tests.sh:
```diff
@@ -61,6 +61,8 @@ run_t3000_ttfabric_tests() {
   #TT_METAL_SLOW_DISPATCH_MODE=1 ./build/test/tt_metal/tt_fabric/fabric_unit_tests --gtest_filter="Fabric2D*Fixture.*"
 
   # these tests cover mux fixture as well
+  TT_METAL_FABRIC_TELEMETRY=1 ./build/test/tt_metal/tt_fabric/fabric_unit_tests --gtest_filter="Fabric2D*Fixture.*"
+  TT_METAL_FABRIC_TELEMETRY=1 ./build/test/tt_metal/tt_fabric/fabric_unit_tests --gtest_filter="Fabric1D*Fixture.*"
```
these set of tests currently run in APC under the t3k fast tests - do we want the telemetry tests to run in APC or can we defer to another pipeline?

aagarwalTT replied:
Also, to be clear, doesn't look like we are actually capturing any telemetry data here, so all we are doing here is making sure the tests run fine with telemetry enabled?

aagarwalTT commented on tt_metal/fabric/hw/inc/edm_fabric/telemetry/fabric_bandwidth_telemetry.hpp:
```diff
@@ -0,0 +1,268 @@
+ * @brief L1 buffer wrapper for storing telemetry data
+ */
+struct L1PerfTelemetrySingleBuffer {
+    volatile tt_l1_ptr uint32_t* buffer_ptr;
```
why do we need this to be volatile?

aagarwalTT commented on tests/ttnn/unit_tests/gtests/ccl/test_fabric_edm_common.hpp:
```diff
@@ -2723,6 +2723,211 @@ tt::tt_fabric::FabricRouterBufferConfig get_edm_buffer_config_helper(FabricTestM
+        auto device_mesh_coord = mesh_device->find_device(device->id());
+
+        for (const auto& eth_logical_core : control_plane.get_active_ethernet_cores(device->id())) {
+            // for (const auto& eth_logical_core : device->get_active_ethernet_cores(true)) {
```
we should also have an inner loop for no. of risc cores to cover BH?

SeanNijjar replied:
no - the L1 is shared and only sender side generates the telemetry

aagarwalTT commented on tests/ttnn/unit_tests/gtests/ccl/test_fabric_edm_common.hpp:
```diff
@@ -2723,6 +2723,211 @@ tt::tt_fabric::FabricRouterBufferConfig get_edm_buffer_config_helper(FabricTestM
+                continue;
+            }
+
+            tt::tt_metal::detail::WriteToDeviceL1(
```
why do we need to explicitly clear out the data though? The kernels can do it on startup?

SeanNijjar replied:
this function is called per test case. The fabric itself is persistent so it doesn't know about the start or end of a given test, so the test code itself does the clearing.

SeanNijjar committed:
make stylistically consistent

SeanNijjar committed:
WIP: add lightweight performance profiling in fabric router

WIP

REVIEW CAREFULLY - SOME DUMPING OF INFO

cleanup

fixes

first shot at (ultra) low-cost fabric in-router BW telemetry

add tests

add visualizations

cleanup a print

correct labels

add missing license

cleanup

cleanup

sq

remove whitespace

SeanNijjar committed:
s

SeanNijjar committed:
add a note about the telemetry flag

SeanNijjar committed:
delete deadcode
